jgrapht-core/src/main/java/org/jgrapht/alg/matching/PathGrowingWeightedMatching.java

```diff
@@ -0,0 +1,444 @@
+        }
+        this.epsilon = epsilon;
+
+        Pair<Double, Set<E>> result;
+        if (useHeuristics) {
+            result = runWithHeuristics(graph);
+        } else {
+            result = run(graph);
+        }
+
+        this.weight = result.first;
+        this.matching = result.second;
+    }
+
+    /**
+     * Get a matching that is a 1/2-approximation of the maximum weighted matching.
+     * 
+     * @return a matching
+     */
+    @Override
+    public Set<E> getMatching()
+    {
+        return Collections.unmodifiableSet(matching);
+    }
+
+    /**
+     * Get the weight of the computed matching.
+     * 
+     * @return the weight of the matching
+     */
+    @Override
+    public double getMatchingWeight()
+    {
+        return weight;
+    }
+
+    // the algorithm (no heuristics)
+    private Pair<Double, Set<E>> run(Graph<V, E> graph)
+    {
+        // lookup all relevant vertices
+        Set<V> visibleVertex = new HashSet<>();
+        for (E e : graph.edgeSet()) {
+            V s = graph.getEdgeSource(e);
+            V t = graph.getEdgeTarget(e);
+            if (!s.equals(t)) {
+                visibleVertex.add(s);
+                visibleVertex.add(t);
+            }
+        }
+
+        // run algorithm
+        Set<E> m1 = new HashSet<>();
+        Set<E> m2 = new HashSet<>();
+        double m1Weight = 0d, m2Weight = 0d;
+        int i = 1;
+        while (!visibleVertex.isEmpty()) {
+            // find vertex arbitrarily
+            V x = visibleVertex.stream().findAny().get();
+
+            // grow path from x
+            while (x != null) {
+                // first heaviest edge incident to vertex x (among visible neighbors)
+                double maxWeight = 0d;
+                E maxWeightedEdge = null;
+                V maxWeightedNeighbor = null;
+                for (E e : graph.edgesOf(x)) {
+                    V other = Graphs.getOppositeVertex(graph, e, x);
+                    if (visibleVertex.contains(other) && !other.equals(x)) {
+                        double curWeight = graph.getEdgeWeight(e);
+                        if (maxWeightedEdge == null || compareWeight(curWeight, maxWeight) > 0) {
+                            maxWeight = curWeight;
+                            maxWeightedEdge = e;
+                            maxWeightedNeighbor = other;
+                        }
+                    }
+                }
+
+                // add it to either m1 or m2, alternating between them
+                if (maxWeightedEdge != null) {
+                    switch (i) {
+                    case 1:
+                        m1.add(maxWeightedEdge);
+                        m1Weight += maxWeight;
+                        break;
+                    case 2:
+                        m2.add(maxWeightedEdge);
+                        m2Weight += maxWeight;
+                        break;
+                    default:
+                        throw new IllegalArgumentException(
```

Comment:
Using IllegalArgumentException like this is confusing.  If assert isn't appropriate, use a plain RuntimeException.